File: src/DynamicProgramming/SubsetSum.java
package DynamicProgramming;

public class SubsetSum {

    // recursive
    // Time Complexity : O(2^n)
    static int countSubsetsRec(int[] arr, int n, int sum) {
        if (n == 0) {
            return (sum == 0) ? 1 : 0;
        }
        return countSubsetsRec(arr, n - 1, sum) + countSubsetsRec(arr, n - 1, sum - arr[n - 1]);
    }

    // Dynamic Programming
    // Time Complexity : O(n * sum)
    // Auxiliary space : O(n * sum)
    static int countSubsets(int[] arr, int n, int sum) {
        int[][] dp = new int[n + 1][sum + 1];

        for (int i = 0; i <= n; i++) {
            dp[i][0] = 1;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= sum; j++) {
                if (j >= arr[i - 1]){
                    dp[i][j] = dp[i - 1][j] + dp[i - 1][j - arr[i - 1]];
                } else {
                    dp[i][j] = dp[i - 1][j];
                }
            }
        }

        return dp[n][sum];
    }

    public static void main(String[] args) {
        int[] arr = {10, 20, 15, 5, 5};
        int sum = 25;
        /*int[] arr = {2, 5, 3};
        int sum = 5;*/
        System.out.println(countSubsetsRec(arr, arr.length, sum));
        System.out.println(countSubsets(arr, arr.length, sum));
    }
}

File: src/DynamicProgramming/DynamicProgramming/SubsetSum.py
# Recursive Approach
# Time Complexity : O(2^n)
def count_subsets_rec(arr, n, sum):
    if n == 0:
        return 1 if sum == 0 else 0
    return count_subsets_rec(arr, n - 1, sum) + count_subsets_rec(arr, n - 1, sum - arr[n - 1])


# Dynamic Programming
# Time Complexity : O(n * sum)
# Auxiliary space : O(n * sum)
def count_subsets(arr, n, sum):
    dp = [[0 for i in range(sum + 1)] for j in range(n + 1)]

    for i in range(n + 1):
        dp[i][0] = 1

    for i in range(1, n + 1):
        for j in range(1, sum + 1):
            if j >= arr[i - 1]:
                dp[i][j] = dp[i - 1][j] + dp[i - 1][j - arr[i - 1]]
            else:
                dp[i][j] = dp[i - 1][j]

    return dp[n][sum]


if __name__ == "__main__":
    arr = [10, 20, 15, 5, 5]
    sm = 25
    print(count_subsets_rec(arr, len(arr), sm))
    print(count_subsets(arr, len(arr), sm))